Update the Digital Asset Standard (#435)
aptos-labs/developer-docs: apps/nextra/pages/en/build/smart-contracts/digital-asset.mdx

title: "Aptos Digital Asset Standard"
---

import { Callout } from 'nextra/components'
import { ThemedImage } from "@components/index";

# Aptos Digital Asset (DA) Standard

The Digital Asset (DA) standard is a modern Non-Fungible Token (NFT) standard for Aptos. NFTs represent unique assets on-chain, and are stored in collections. These NFTs can be customized to later be transferred, soulbound, burned, mutated, or customized via your own smart contracts.

This standard replaces the legacy [Aptos Token Standard](aptos-token.mdx). The most important improvements to note are:

> 1. Tokens can be easily extended since they are implemented using Move [Objects](objects.mdx).
> 2. You can now directly transfer NFTs without the recipient “opting-in” on-chain.
> 3. NFTs can own other NFTs for easy composability. 

Reviewer: ```suggestion
| **Improvement** | **Description** |
|-----------------|-----------------|
| Token Extension | Tokens can be easily extended since they are implemented using Move [Objects](objects.mdx). |
| Direct NFT Transfer | You can now directly transfer NFTs without the recipient “opting-in” on-chain. |
| NFT Composability | NFTs can own other NFTs for easy composability. |

```

If you want a simple way to mint NFTs without the ability to customize or extend their functionality, you can use the `aptos_token` module which implements the DA standard (see the section on how to use it below). 

<Callout type="info">
Note that all Digital Asset modules are deployed at the reserved address `0x4`.
</Callout>

## Using the Digital Asset Standard

This standard is implemented with two Objects:

1. `Collection`s - A set of NFTs with a name and a bit of context for the group.
2. `Token`s - Digital assets which represent unique assets. These are often used to represent NFTs and usually use a `uri` link to point to more info about the asset (ex. a link to an image, video, etc.).

<ThemedImage
alt="Digital Asset token and collection relationship"
sources={{
    light: '/docs/digital-asset-light.svg',
    dark: '/docs/digital-asset-dark.svg',
  }}
/>

All `Token`s are required to have a reference to a parent `Collection`, but the parent `Collection` does **not** own the `Token`. Newly minted `Token`s are usually owned by the creator. From there, they can be transferred to other accounts.

> ### Creating a `Collection`

Reviewer: ```suggestion
## Creating a `Collection`
```

There are two ways to create a `Collection` depending on whether you want there to be a maximum supply of NFTs it can hold:

> 1. For a fixed maximum supply you can use `collection::create_fixed_collection`:

Reviewer: ```suggestion
### Fixed Maximum Supply

The maximum supply of the collection is fixed to a set number less than __MAX_BOUNDS__ and cannot be changed in the future. Doing this requires `collection::create_fixed_collection`
```

Author: I'm wary of including __MAX_BOUNDS__ without indicating where to find that value or what the number is. Do you happen to know what it is? (I couldn't find it from searching around). 

Updated the header and text otherwise :)

> ```move filename="Example.move"

Reviewer: ```suggestion
```move filename="example.move"
```

use aptos_token_objects::collection;

public entry fun create_collection(creator: &signer) {

> let max_supply = 1000;

Reviewer: ```suggestion
    let max_supply = 1000;
```

    // Maximum supply cannot be changed after collection creation
    collection::create_fixed_collection(
        creator,
        "My Collection Description",
        max_supply,
        "My Collection",
        royalty,
        "https://mycollection.com",
    );
}
```

> 1. For a `Collection` with unlimited supply you can use `collection::create_unlimited_collection`:

Reviewer: ```suggestion
### Unlimited Supply

The supply of the collection is unlimited. This requires `collection::create_unlimited_collection`:
```

> ```move filename="Example.move"

Reviewer: ```suggestion
```move filename="example.move"
```

public entry fun create_collection(creator: &signer) {
    collection::create_unlimited_collection(
        creator,
        "My Collection Description",
        "My Collection",
        royalty,
        "https://mycollection.com",
    );
}
```

<Callout type="warning">
A `Collection`'s maximum supply cannot be changed after creation.
</Callout>

> `Collection`s have the following fields:
> 
> - Description - an optional string smaller than 2048 characters (modifiable with a `MutatorRef`).
> - Name - a required string to identify the `Collection`. The name must be unique within each account. That means a single creator account cannot create more than one `Collection` with the same name.
> - Royalty - an optional number indicating what % of the sale price goes to the creator of the `Collection`. This can be changed with a `MutatorRef` generated by the [Royalty module](https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/framework/aptos-token-objects/sources/royalty.move) (an extension for the DA standard. See example usage in [`aptos_token.move`](https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/framework/aptos-token-objects/sources/aptos_token.move))
> - URI length - an optional string that is smaller than 512 characters which links to relevant content for the `Collection` (modifiable with a `MutatorRef`).

Reviewer: Let's move this to right below the `## Collections` header

```suggestion
| **Field** | **Description** |
|-----------|-----------------|
| **Description** | An optional string smaller than 2048 characters (modifiable with a `MutatorRef`). |
| **Name** | A required string to identify the `Collection`. The name must be unique within each account. That means a single creator account cannot create more than one `Collection` with the same name. |
| **Royalty** | An optional number indicating what % of the sale price goes to the creator of the `Collection`. This can be changed with a `MutatorRef` generated by the [Royalty module](https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/framework/aptos-token-objects/sources/royalty.move) (an extension for the DA standard. See example usage in [`aptos_token.move`](https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/framework/aptos-token-objects/sources/aptos_token.move)). |
| **URI length** | An optional string that is smaller than 512 characters which links to relevant content for the `Collection` (modifiable with a `MutatorRef`). |

```

Since each `Collection` is a [Move Object](objects.mdx), you can customize it by generating permissions called `Ref`s. Each `Ref` allows you to modify an aspect of the Object later on. Beyond the normal [Object Refs](objects/creating-objects.mdx), `Collection`s can also get a `MutatorRef` by calling `get_mutator_ref` like so:

> ```move filename="Example.move"

Reviewer: ```suggestion
```move filename="example.move"
```

public entry fun create_collection(creator: &signer) {
    let collection_constructor_ref = &collection::create_unlimited_collection(
        creator,
        "My Collection Description",
        "My Collection",
        royalty,
        "https://mycollection.com",
    );
    let mutator_ref = collection::get_mutator_ref(collection_constructor_ref);
    // Store the mutator ref somewhere safe
}
```

<Callout type="warning">
Refs **must** be generated at creation time of an Object. The `ConstructorRef` used to generate other `Ref`s expires as soon as the transaction to create the Object is finished.
</Callout>

You can further customize your `Collection` by adding more resources or functionalities via smart contract. For example, a `Collection` can track when it was created in order to limit when `Token`s can be minted like so:

> ```move filename="Example.move"

Reviewer: ```suggestion
```move filename="example.move"
```

struct MyCollectionMetadata has key {
    creation_timestamp_secs: u64,
}

public entry fun create_collection(creator: &signer) {
    // Constructor ref is a non-storable struct returned when creating a new object.
    // It can generate an object signer to add resources to the collection object.
    let collection_constructor_ref = &collection::create_unlimited_collection(
        creator,
        "My Collection Description",
        "My Collection",
        royalty,
        "https://mycollection.com",
    );
    // Constructor ref can be exchanged for signer to add resources to the collection object.
    let collection_signer = &object::generate_signer(collection_constructor_ref);
    move_to(collection_signer, MyCollectionMetadata { creation_timestamp_secs: timestamp::now_seconds() } })
}
```

> ## Creating Tokens

Reviewer: Let's add a header `## Tokens` and then make `### Creating Tokens` a subheader

`Token`s have similar fields to `Collection`s: 

> - Description - an optional string smaller than 2048 characters (modifiable with a `MutatorRef`).
> - Name - a required string to identify the `Collection` that is unique within each `Collection`. This means a single `Collection` account cannot have more than one `Token` with the same name.
> - Royalty - an optional number indicating what % of the sale price goes to the creator of the `Collection`. This can be changed with a `MutatorRef` generated by the [Royalty module](https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/framework/aptos-token-objects/sources/royalty.move) (an extension for the DA standard. See example usage in [`aptos_token.move`](https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/framework/aptos-token-objects/sources/aptos_token.move)). Usually royalty is set on collections, but setting it on `Token`s allows the individual `Token` to have a custom royalty amount.
> - URI length - an optional string that is smaller than 512 characters which links to relevant content for the `Collection` (modifiable with a `MutatorRef`).

Reviewer: Put this under `## Tokens`

```suggestion
| **Field** | **Description** |
|-----------|-----------------|
| **Description** | An optional string smaller than 2048 characters (modifiable with a `MutatorRef`). |
| **Name** | A required string to identify the `Collection` that is unique within each `Collection`. This means a single `Collection` account cannot have more than one `Token` with the same name. |
| **Royalty** | An optional number indicating what % of the sale price goes to the creator of the `Collection`. This can be changed with a `MutatorRef` generated by the [Royalty module](https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/framework/aptos-token-objects/sources/royalty.move) (an extension for the DA standard. See example usage in [`aptos_token.move`](https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/framework/aptos-token-objects/sources/aptos_token.move)). Usually royalty is set on collections, but setting it on `Token`s allows the individual `Token` to have a custom royalty amount. |
| **URI length** | An optional string that is smaller than 512 characters which links to relevant content for the `Collection` (modifiable with a `MutatorRef`). |

```

There are a few ways to create a `Token`:

1. Named tokens. These use the name of the `Token` to generate a named Object. This makes it easy to find the address for the token if you know the token and `Collection` name, but named Objects are not deletable. Trying to delete the a named token will only delete the data, not the Object itself.
    

> ```move filename="Example.move"

Reviewer: ```suggestion
```move filename="example.move"
```

use aptos_token_objects::token;

public entry fun mint_token(creator: &signer) {
    token::create_named_token(
        creator,
        "Collection Name",
        "Description",
        "Token Name",
        royalty,
        "https://mycollection.com/my-named-token.jpeg",
    );
}
```
    
<Callout type="info">
You can derive the address for named tokens by:
1. Concatenating the creator address, collection name and token name.
2. Doing a sha256 hash of that new string.
</Callout>
    
2. “Unnamed” tokens. These create unnamed *Objects* (which **are** deletable) but still have a `Token` name. Because the Object address is not deterministic, you must use an Indexer to find the address for them.
    

> ```move filename="Example.move"

Reviewer: ```suggestion
```move filename="example.move"
```

use aptos_token_objects::token;

public entry fun mint_token(creator: &signer) {
    token::create(
        creator,
        "Collection Name",
        "Description",
        "Token Name",
        royalty,
        "https://mycollection.com/my-named-token.jpeg",
    );
}
```
    

> You can find the addresses of your recently created “unnamed” `Token`s by using the [Aptos Indexer](../indexer/api/labs-hosted.mdx) with queries like the following:

Reviewer: Add header above this like `## Indexer Token Queries`

1. Looking up the collection id by using your account address and the name of the `Collection`. 

```bash filename="gRPC"
query GetCollectionIdByName($creatorAddress: String!, $collectionName: String!) {
  current_collections_v2(
    where: {creator_address: {_eq: $creatorAddress}, collection_name: {_eq: $collectionName}}
  ) {
    collection_name
    collection_id
  }

> }

Reviewer: We should embed the indexer graphql component here

  "collectionName": "Horse Emojis"
}
```

1. Then look up the address (`token_data_id`) of the `Token` by using the `collection_id` (from above) and the name of the token:
    
```bash filename="gRPC"
query GetTokenID($collection_id: String!, $token_name: String!) {
  current_token_datas_v2(
    where: {collection_id: {_eq: $collection_id}, token_name: {_eq: $token_name}}
  ) {
    token_name
    description
    token_data_id
  }
}
```
    
```bash filename="gRPC"
Query Variables:
{
  "collection_id": "0x158dc6481aec80004d0f21a8994757e3c8203fc37b0f508be2a5364ad99e649e",
  "token_name": "Horse Emojis #127"
}
```
    
<Callout type="info">
In general, using unnamed tokens give you the most flexibility because the Object can be deleted later, but named tokens simplify looking up addresses.
</Callout>

### Customizing Tokens

`Token`s can be extended either by adding additional resources (since they are an Object) or using `Ref`s to modify the Object.

1. Transferring a `Token` can be done by calling [`object::transfer`](https://aptos.dev/reference/move/?branch=mainnet&page=aptos-framework/doc/object.md#0x1_object_transfer).
    

> ```move filename="Example.move"

Reviewer: ```suggestion
```move filename="example.move"
```

public entry fun transfer<T: key>(owner: &signer, object: object::Object<T>, to: address)
```
    
2. Burning / deleting a `Token` requires storing a `BurnRef` with `token::generate_burn_ref`, then calling `token::burn`.
    

> ```move filename="Example.move"

Reviewer: ```suggestion
```move filename="example.move"
```

public entry fun mint_token(creator: &signer) {
    let token_constructor_ref = &token::create(
        creator,
        "My Collection",
        "My named Token description",
        "My named token",
        royalty,
        "https://mycollection.com/my-named-token.jpeg",
    );
    let burn_ref = token::generate_burn_ref(token_constructor_ref);
    // Store the burn ref somewhere safe
}

public entry fun burn_token(token: Object<Token>) {
    // Remove all custom data from the token object.
    let token_address = object::object_address(&token);
    let CustomData { ... } = move_from<CustomData>(token_address);

    // Retrieve the burn ref from storage
    let burn_ref = ...;
    token::burn(burn_ref);
}
```
    
<Callout type="warning">
If any custom resources were moved onto the Token, those must be removed / deleted first before`token::burn` can delete the Token. For named tokens which cannot be deleted, `token::burn` will For named Tokens `token::burn` will remove all Token content instead.
</Callout>
    
3. Mutating a `Token`’s `URI` or `description` requires a `MutatorRef`.
    

> ```move filename="Example.move"

Reviewer: ```suggestion
```move filename="example.move"
```

    

## Aptos Token

For NFT creators who want to avoid writing their own logic for how your NFT should work, you can use the [`aptos_token`](https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/framework/aptos-token-objects/sources/aptos_token.move) module to mint an NFT. This module is already deployed at `0x4` and allows you to:

1. Mint a `Token` you can transfer with royalties.
2. Mint a soulbound `Token`.
3. Manage the resources your NFT has.

See the `aptos_token` [reference docs](https://aptos.dev/reference/move/?branch=mainnet&page=aptos-token-objects/doc/aptos_token.md) for all the helper functions you can use.

<Callout type="warning">
The main drawback of using the `aptos_token` module is that the Tokens are not extensible (the `mint` function does not return a `ConstructorRef`).
</Callout>

### Minting with `aptos_token`

Minting a `Token` using `aptos_token` requires the same parameters as any token that implements the DA standard. In addition though, the `aptos_token` module allows you to specificy a property map of key/value pairs for any other properties your specific NFT may require. 

You can mint your `Token` by calling `aptos_token::mint` like so: 

> ```move filename="Example.move"

Reviewer: ```suggestion
```move filename="example.move"
```

title: "Aptos Digital Asset Standard"
---

import { Callout } from 'nextra/components'
import { ThemedImage, GraphQLEditor } from "@components/index";

# Aptos Digital Asset (DA) Standard

The Digital Asset (DA) standard is a modern Non-Fungible Token (NFT) standard for Aptos. NFTs represent unique assets on-chain, and are stored in collections. These NFTs can be customized to later be transferred, soulbound, burned, mutated, or customized via your own smart contracts.

This standard replaces the legacy [Aptos Token Standard](aptos-token.mdx). The most important improvements to note are:

| **Improvement** | **Description** |
|-----------------|-----------------|

> | Token Extension | Tokens can be easily extended since they are implemented using Move [Objects](objects.mdx). |
> | Direct NFT Transfer | You can now directly transfer NFTs without the recipient “opting-in” on-chain. |
> | NFT Composability | NFTs can own other NFTs for easy composability. |

Reviewer: ```suggestion
| **Token Extension** | Tokens can be easily extended since they are implemented using Move [Objects](objects.mdx). |
| **Direct NFT Transfer** | You can now directly transfer NFTs without the recipient “opting-in” on-chain. |
| **NFT Composability** | NFTs can own other NFTs for easy composability. |
```

| **Royalty** | An optional number indicating what % of the sale price goes to the creator of the `Collection`. This can be changed with a `MutatorRef` generated by the [Royalty module](https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/framework/aptos-token-objects/sources/royalty.move). The Royalty module is an extension for the DA standard. See example usage in [`aptos_token.move`](https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/framework/aptos-token-objects/sources/aptos_token.move). |
| **URI length** | An optional string that is smaller than 512 characters which links to relevant content for the `Collection` (modifiable with a `MutatorRef`). |

### Creating a `Collection`

There are two ways to create a `Collection` depending on whether you want there to be a maximum supply of `Token`s it can hold.

#### Fixed Maximum Supply

To make a `Collection` with a fixed supply you can use `collection::create_fixed_collection` like so:

```move filename="example.move"
use aptos_token_objects::collection;

public entry fun create_collection(creator: &signer) {
    let max_supply = 1000;

>     

Reviewer: Need to define a `royalty` variable

    // Maximum supply cannot be changed after collection creation
    collection::create_fixed_collection(
        creator,
        "My Collection Description",
        max_supply,
        "My Collection",
        royalty,
        "https://mycollection.com",
    );
}
```
#### Unlimited Supply

To create a `Collection` with unlimited supply you can use `collection::create_unlimited_collection`:

```move filename="example.move"
public entry fun create_collection(creator: &signer) {
    collection::create_unlimited_collection(
        creator,
        "My Collection Description",
        "My Collection",

>         royalty,

Reviewer: Need to define a `royalty` variable

        "https://mycollection.com",
    );
}
```

<Callout type="warning">
A `Collection`'s maximum supply cannot be changed after creation.
</Callout>

### Customizing a `Collection`

Since each `Collection` is a [Move Object](objects.mdx), you can customize it by generating permissions called `Ref`s. Each `Ref` allows you to modify an aspect of the Object later on. Beyond the normal [Object Refs](objects/creating-objects.mdx), `Collection`s can also get a `MutatorRef` by calling `get_mutator_ref` like so:

```move filename="example.move"
public entry fun create_collection(creator: &signer) {
    let collection_constructor_ref = &collection::create_unlimited_collection(
        creator,
        "My Collection Description",
        "My Collection",

>         royalty,

Reviewer: Need to define a `royalty` variable

        "https://mycollection.com",
    );
    let mutator_ref = collection::get_mutator_ref(collection_constructor_ref);
    // Store the mutator ref somewhere safe
}
```

<Callout type="warning">
Refs **must** be generated at creation time of an Object. The `ConstructorRef` used to generate other `Ref`s expires as soon as the transaction to create the Object is finished.
</Callout>

You can further customize your `Collection` by adding more resources or functionalities via smart contract. For example, a `Collection` can track when it was created in order to limit when `Token`s can be minted like so:

```move filename="example.move"
struct MyCollectionMetadata has key {
    creation_timestamp_secs: u64,
}

public entry fun create_collection(creator: &signer) {
    // Constructor ref is a non-storable struct returned when creating a new object.
    // It can generate an object signer to add resources to the collection object.
    let collection_constructor_ref = &collection::create_unlimited_collection(
        creator,
        "My Collection Description",
        "My Collection",

>         royalty,

Reviewer: Need to define a `royalty` variable

| **URI length** | An optional string that is smaller than 512 characters which links to relevant content for the `Collection` (modifiable with a `MutatorRef`). |

### Creating Tokens 

There are a few ways to create a `Token`:

1. Named tokens. These use the name of the `Token` to generate a named Object. This makes it easy to find the address for the token if you know the token and `Collection` name, but named Objects are not deletable. Trying to delete the a named token will only delete the data, not the Object itself.
    
```move filename="example.move"
use aptos_token_objects::token;

public entry fun mint_token(creator: &signer) {
    token::create_named_token(
        creator,
        "Collection Name",
        "Description",
        "Token Name",

>         royalty,

Reviewer: Need to define a `royalty` variable

        "https://mycollection.com/my-named-token.jpeg",
    );
}
```
    
<Callout type="info">
You can derive the address for named tokens by:
1. Concatenating the creator address, collection name and token name.
2. Doing a sha256 hash of that new string.
</Callout>
    
2. “Unnamed” tokens. These create unnamed *Objects* (which **are** deletable) but still have a `Token` name. Because the Object address is not deterministic, you must use an Indexer to find the address for them.
    
```move filename="example.move"
use aptos_token_objects::token;

public entry fun mint_token(creator: &signer) {
    token::create(
        creator,
        "Collection Name",
        "Description",
        "Token Name",

>         royalty,

Reviewer: Need to define a `royalty` variable

    token_data_id
  }
}`}
  variables={`{
  "collection_id": "0x158dc6481aec80004d0f21a8994757e3c8203fc37b0f508be2a5364ad99e649e",
  "token_name": "Horse Emojis #127"
}`}
/>
    
<Callout type="info">
In general, using unnamed tokens give you the most flexibility because the Object can be deleted later, but named tokens simplify looking up addresses.
</Callout>

### Customizing Tokens

`Token`s can be extended either by adding additional resources (since they are an Object) or using `Ref`s to modify the Object.

1. Transferring a `Token` can be done by calling [`object::transfer`](https://aptos.dev/reference/move/?branch=mainnet&page=aptos-framework/doc/object.md#0x1_object_transfer).
    
```move filename="example.move"
public entry fun transfer<T: key>(owner: &signer, object: object::Object<T>, to: address)
```
    
2. Burning / deleting a `Token` requires storing a `BurnRef` with `token::generate_burn_ref`, then calling `token::burn`.
    
```move filename="example.move"
public entry fun mint_token(creator: &signer) {
    let token_constructor_ref = &token::create(
        creator,
        "My Collection",
        "My named Token description",
        "My named token",

>         royalty,

Reviewer: Need to define a `royalty` variable

    // Store the burn ref somewhere safe
}

public entry fun burn_token(token: Object<Token>) {
    // Remove all custom data from the token object.
    let token_address = object::object_address(&token);
    let CustomData { ... } = move_from<CustomData>(token_address);

    // Retrieve the burn ref from storage
    let burn_ref = ...;
    token::burn(burn_ref);
}
```
    
<Callout type="warning">
If any custom resources were moved onto the Token, those must be removed / deleted first before`token::burn` can delete the Token. For named tokens which cannot be deleted, `token::burn` will For named Tokens `token::burn` will remove all Token content instead.
</Callout>
    
3. Mutating a `Token`’s `URI` or `description` requires a `MutatorRef`.
    
```move filename="example.move"
public entry fun mint_token(creator: &signer) {
    // Constructor ref is a non-storable struct returned when creating a new object.
    // It can be exchanged for signer to add resources to the token object.
    let token_constructor_ref = &token::create(
        creator,
        "My Collection",
        "My named Token description",
        "My named token",

>         royalty,

Reviewer: Need to define a `royalty` variable

    property_keys: vector<String>,
    property_types: vector<String>,
    property_values: vector<vector<u8>>,
) acquires AptosCollection, AptosToken
```

> To mint a soul bound `Token`, you can call [`aptos_token::mint_soul_bound`](https://aptos.dev/reference/move/?branch=mainnet&page=aptos-token-objects/doc/aptos_token.md#0x4_aptos_token_mint_soul_bound) instead:

Reviewer: Add a header for `### Soulbound Token`

> ```move filename="Example.move"

Reviewer: ```suggestion
```move filename="example.move"
```
